contracts/cw1155-base/src/contract.rs

```diff
@@ -0,0 +1,1239 @@
+        );
+
+        // batch query
+        assert_eq!(
+            query(
+                deps.as_ref(),
+                mock_env(),
+                Cw1155QueryMsg::BatchBalance {
+                    owner: user1.clone(),
+                    token_ids: vec![token1.clone(), token2.clone(), token3.clone()],
+                }
+            ),
+            to_binary(&BatchBalanceResponse {
+                balances: vec![1u64.into(), 1u64.into(), 1u64.into()]
+            }),
+        );
+
+        // user1 revoke approval
+        execute(
+            deps.as_mut(),
+            mock_env(),
+            mock_info(user1.clone(), &[]),
+            Cw1155ExecuteMsg::RevokeAll {
+                operator: minter.clone(),
+            },
+        )
+        .unwrap();
+
+        // query approval status
+        assert_eq!(
+            query(
+                deps.as_ref(),
+                mock_env(),
+                Cw1155QueryMsg::IsApprovedForAll {
+                    owner: user1.clone(),
+                    operator: minter.clone(),
+                }
+            ),
+            to_binary(&IsApprovedForAllResponse { approved: false }),
+        );
+
+        // tranfer without approval
+        assert!(matches!(
+            execute(
+                deps.as_mut(),
+                mock_env(),
+                mock_info(minter.clone(), &[]),
+                Cw1155ExecuteMsg::SendFrom {
+                    from: user1.clone(),
+                    to: user2.clone(),
+                    token_id: token1.clone(),
+                    value: 1u64.into(),
+                    msg: None,
+                },
+            ),
+            Err(ContractError::Unauthorized {})
+        ));
+
+        // burn token1
+        assert_eq!(
+            execute(
+                deps.as_mut(),
+                mock_env(),
+                mock_info(user1.clone(), &[]),
+                Cw1155ExecuteMsg::Burn {
+                    from: user1.clone(),
+                    token_id: token1.clone(),
+                    value: 1u64.into(),
+                }
+            )
+            .unwrap(),
+            Response {
+                attributes: vec![
+                    attr("action", "transfer"),
+                    attr("token_id", &token1),
+                    attr("amount", 1u64),
+                    attr("from", &user1),
+                ],
+                ..Response::default()
+            }
+        );
+
+        // burn them all
+        assert_eq!(
+            execute(
+                deps.as_mut(),
+                mock_env(),
+                mock_info(user1.clone(), &[]),
+                Cw1155ExecuteMsg::BatchBurn {
+                    from: user1.clone(),
+                    batch: vec![(token2.clone(), 1u64.into()), (token3.clone(), 1u64.into())]
+                }
+            )
+            .unwrap(),
+            Response {
+                attributes: vec![
+                    attr("action", "transfer"),
```

Comment:
This is interesting. I am not sure how this ends up when it comes out through tendermint (I think it does some "collapsing" of events).

It totally makes sense. I wonder if you had a wish on a cleaner way to encode these or pass them through to the client? Although I guess just as it is is perfect for searching.

Reply:
I'm not sure how it's going to be indexed by tendermint though, need to look into details in tendermint I guess.

Reply:
Yeah, they should all match